### ISSUE TITLE
"TypeError: Cannot call method 'match' of undefined" on message receipt

### ISSUE THREAD

**Issue by kfarwell** [hzoo/TwitchPlaysX#4]
server connects to the IRC channel fine, crashes when someone sends a command matched by regex (command ls sent by twitch_i in this example)
```
% node ./server.js
twitch_i                                                           ls

/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:759
                    throw err;
                          ^
TypeError: Cannot call method 'match' of undefined
    at Object.sendKey (/home/kuko/twitch/TwitchPlaysX/app/keyHandler.js:28:17)
    at Client.<anonymous> (/home/kuko/twitch/TwitchPlaysX/app/server.js:38:24)
    at Client.emit (events.js:106:17)
    at Client.<anonymous> (/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:520:26)
    at Client.emit (events.js:95:17)
    at iterator (/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:756:22)
    at Array.forEach (native)
    at Socket.<anonymous> (/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:752:15)
    at Socket.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:764:14)
% 
```

config.js
```
//list of commands to filter for
var commands = [
    'left', 'right', 'up', 'down', 'start', 'select', 'a', 'b', 'democracy', 'anarchy'
];

var ircConfig = {
    //either 'windows' or 'other'
    os: 'other',

    //program
    //title of the window of the program
    //ex: Desmume for DS
    programName: 'virt-manager',
    // programName: 'Desmume',

    //IRC
    //ex: irc.twitch.tv or 199.9.252.26
    server: 'irc.twitch.tv',
    //ex: your twitch username
    nick: 'removed',
    //ex: your twitch username
    userName: 'removed',
    //oauth token from www.twitchapps.com/tmi
    password: 'oauth:removed',
    //ex: #twitchplayspokemon
    channel: '#removed',
    //if you want to join more than one channel
    // CLARIFICATION: Main channel has to be in here too!
    channelList: ['#removed'],

    //other IRC
    //default
    port: 6667,
    //needed for password
    // Twitch uses server pass, no sasl
    sasl: false,
    //if you want to get all messages in real-time keep false
    floodProtection: false,
    //only needed if true - in milliseconds
    floodProtectionDelay: 100,

    //config
    //if you want to send keys to a program
    sendKey: true,
    //if you want to print usernames/commands like in twitchplayspokemon
    printToConsole: true,

    //commands
    //example of regex
    //don't need to modify if you need anarchy mode
    //matches for any of the words in 'commands'
    regexCommands: new RegExp('^(' + '.*' + ')$', 'i'),
    //if you need to filter the commands sent to the program
    //ex: democracy/anarchy since they don't affect the program itself
    filteredCommands: [
    ],
    //if you want to prevent people from using from command too often
    throttledCommands: [
    ],
    //maximum characters to show for a person's name in the console log
    maxCharName: 8,
    //maximum characters to show for a command in the console log
    //ex: left => left since only 4 char, democracy => democra
    maxCharCommand: 60,
    //timeToWait between keypresses if you want to limit a certain button
    //ex: you can limit 'start' so it's only used every 10 sec
    timeToWait: 100,
    //keymap for program
    //need to set this for your program correcly (vba,desmume,etc)
    keymap: {
    },
    //linux: delay between each possible keypress in ms (can't be too fast)
    //if you want to change delay for windows - change key.py
    delay: 100
};

module.exports = ircConfig;
```

**Comment by hzoo** [comment]
Looking into this now.. I could redo a bunch of it from the changes in https://github.com/hzoo/tppstats (which was based on this originally) since there is probably a lot of extraneous stuff.

Looks like the error is from `keyHandler.js line 28`. key is undefined

```js
        key = config.keymap[command];
        //throttle certain commands (not individually though)
        if (key.match(regexThrottle)) {
```

**Comment by hzoo** [comment]
Ok I think it's because in your config, the `keymap` property is an empty object. Thus config.keymap[command] would be undefined.

**Comment by hzoo** [comment]
Ok hopefully it works now? `npm install` again after updating/recloning.

Create a config.json file at the root of the folder.

```
{
    "TWITCH_OAUTH": "OAUTH_HERE",
    "TWITCH_USERNAME": "TWITCH_NAME_HERE",
    "TWITCH_CHANNEL": "#CHANNEL_HERE",
    "CONFIG_OS": "other",
    "CONFIG_PROGRAM_NAME": "virt-manager",
    "CONFIG_MAX_CHAR_NAME": 8,
    "CONFIG_MAX_CHAR_COMMAND": 60,
    "CONFIG_SEND_KEY": true,
    "CONFIG_FILTERED_COMMANDS": [],
    "CONFIG_THROTTLED_COMMANDS": []
}
```

then run `npm start`

**Comment by kfarwell** [comment]
> Ok hopefully it works now? npm install again after updating/recloning.

> Create a config.json file at the root of the folder.

> ...

> then run npm start

Did this, now it hangs connecting and doesn't receive any commands from IRC.
```
% npm start

> TwitchPlaysX@0.0.1 start /home/kuko/twitch/TwitchPlaysX
> node ./app/server

Connecting...

```

**Comment by hzoo** [comment]
Ok it was working for me before. What I did was reset my oauth key and it worked.

**Comment by kfarwell** [comment]
I did notice the password reset so I did test with a new oauth key. I couldn't login to IRC with sic either prior to the reset but can now. TwitchPlaysX hung at connecting before and after the password reset and new oauth key.

I was on node 0.10.33 and just upgraded to 0.10.37 with no change. I'm on Trisquel 7.0 (Ubuntu 14.04-based).

**Comment by hzoo** [comment]
Is it actually hanging? It might be that the messages are slow or don't match the commands. Maybe try it on another channel like 

```js
"TWITCH_CHANNEL": "#twitchplayspokemon",
```

or try `console.log(message)` before line 26 https://github.com/hzoo/TwitchPlaysX/blob/master/app/server.js#L26 to see if any messages are coming through.

**Comment by kfarwell** [comment]
You're right, it wasn't matching any commands. 'a' works. I forgot about the regex change I made to `app/config.js` before recloning in an attempt to filter in all commands: `regexCommands: new RegExp('^(.*)$', 'i'),`. I'm guessing my regex is wrong, I just replaced `' + config.commands.join('|') + '` with `.*`. I added my regex change to `app/server.js:23` since the option was removed from `app/config.js`:
```
var commandRegex = config.regexCommands ||
new RegExp('^(.*)$', 'i');
```

I now get the same error:
```
% npm start
> TwitchPlaysX@0.0.1 start /home/kuko/twitch/TwitchPlaysX
> node ./app/server
Connecting...
twitch_i                                                            a
atwitch_i                                                        start
stwitch_i                                                           ls
/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:759
                        throw err;
                              ^
TypeError: Cannot call method 'match' of undefined
    at Object.sendKey (/home/kuko/twitch/TwitchPlaysX/app/keyHandler.js:35:17)
    at Client.<anonymous> (/home/kuko/twitch/TwitchPlaysX/app/server.js:41:24)
    at Client.emit (events.js:106:17)
    at Client.<anonymous> (/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:520:26)
    at Client.emit (events.js:95:17)
    at iterator (/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:756:26)
    at Array.forEach (native)
    at Socket.<anonymous> (/home/kuko/twitch/TwitchPlaysX/node_modules/irc/lib/irc.js:751:15)
    at Socket.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:765:14)
npm ERR! TwitchPlaysX@0.0.1 start: `node ./app/server`
npm ERR! Exit status 8
npm ERR! 
npm ERR! Failed at the TwitchPlaysX@0.0.1 start script.
npm ERR! This is most likely a problem with the TwitchPlaysX package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./app/server
npm ERR! You can get their info via:
npm ERR!     npm owner ls TwitchPlaysX
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 3.13.0-44-generic
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! cwd /home/kuko/twitch/TwitchPlaysX
npm ERR! node -v v0.10.37
npm ERR! npm -v 1.4.28
npm ERR! code ELIFECYCLE
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/kuko/twitch/TwitchPlaysX/npm-debug.log
npm ERR! not ok code 0
% cat npm-debug.log
ascat: No such file or directory
% cat npm-debug.log
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@1.4.28
3 info using node@v0.10.37
4 verbose node symlink /usr/bin/node
5 verbose run-script [ 'prestart', 'start', 'poststart' ]
6 info prestart TwitchPlaysX@0.0.1
7 info start TwitchPlaysX@0.0.1
8 verbose unsafe-perm in lifecycle true
9 info TwitchPlaysX@0.0.1 Failed to exec start script
10 error TwitchPlaysX@0.0.1 start: `node ./app/server`
10 error Exit status 8
11 error Failed at the TwitchPlaysX@0.0.1 start script.
11 error This is most likely a problem with the TwitchPlaysX package,
11 error not with npm itself.
11 error Tell the author that this fails on your system:
11 error     node ./app/server
11 error You can get their info via:
11 error     npm owner ls TwitchPlaysX
11 error There is likely additional logging output above.
12 error System Linux 3.13.0-44-generic
13 error command "/usr/bin/node" "/usr/bin/npm" "start"
14 error cwd /home/kuko/twitch/TwitchPlaysX
15 error node -v v0.10.37
16 error npm -v 1.4.28
17 error code ELIFECYCLE
18 verbose exit [ 1, true ]
```

It also looks like matched commands are being sent to the active st term, not virt-manager as defined by `CONFIG_PROGRAM_NAME`. My window manager doesn't show window titles anywhere, but xlsclients says:

```
% xlsclients -l
Window 0x1200001:
  Machine:  yoka
  Name:  IceCat
  Icon Name:  icecat
  Command:  icecat
  Instance/Class:  icecat/Icecat
Window 0x1000001:
  Machine:  yoka
  Name:  virt-manager
  Icon Name:  virt-manager
  Command:  virt-manager
  Instance/Class:  virt-manager/Virt-manager
```
I'm not sure that `Name` is the property I'm looking for; I think IceCat (Firefox) normally displays a lot more info in the title bar. Is there a more reliable way to determine X window names other than getting a different window manager?

**Comment by hzoo** [comment]
Ok error is on https://github.com/hzoo/TwitchPlaysX/blob/master/app/keyHandler.js#L35. `key is undefined meaning `defaultKeyMap[command] is undefined` which is https://github.com/hzoo/TwitchPlaysX/blob/master/app/keyHandler.js#L22. I think I removed config.keymap so it should default to that object. I think I'm just going to remove the throttle thing since it's not a big deal anyway.

I didn't explore enough with xdotool and stuff like that to figure out how to send it to the correct program - it definitely was frustrating to do.

**Comment by kfarwell** [comment]
No longer crashes, and now sends commands to virt-manager (played around with `xdotool search` until I found the right window). Thanks for all the help!

In order to accept multi-character commands and submit them to a command line, I added spaces between each character in `key` when passing it to xdotool and sent ' Return':
```
exec('xdotool key --window ' + windowID + ' --delay ' + config.delay + ' ' + key.split("").join(" ") + ' Return');
```
I'd like to skip converting `key` to lower-case too, but forget where that happens.
